pkg/operatorstatus/operator_status.go

```diff
@@ -198,13 +200,133 @@ func NewClusterOperatorStatusCondition(conditionType configv1.ClusterStatusCondi
 	}
 }
 
-func printOperandVersions(versions []configv1.OperandVersion) string {
-	versionsOutput := []string{}
-	for _, operand := range versions {
-		versionsOutput = append(versionsOutput, fmt.Sprintf("%s: %s", operand.Name, operand.Version))
+func unsupportedConditions() []configv1.ClusterOperatorStatusCondition {
+	return []configv1.ClusterOperatorStatusCondition{
+		NewClusterOperatorStatusCondition(configv1.OperatorAvailable, configv1.ConditionTrue, ReasonAsExpected, "Cluster API is not yet implemented on this platform"),
+		NewClusterOperatorStatusCondition(configv1.OperatorProgressing, configv1.ConditionFalse, ReasonAsExpected, ""),
+		NewClusterOperatorStatusCondition(configv1.OperatorDegraded, configv1.ConditionFalse, ReasonAsExpected, ""),
+		NewClusterOperatorStatusCondition(configv1.OperatorUpgradeable, configv1.ConditionTrue, ReasonAsExpected, ""),
+	}
+}
+
+func aggregatedStatusConditions(co *configv1.ClusterOperator, releaseVersion string, currentFeatureGates featuregates.FeatureGate) []configv1.ClusterOperatorStatusCondition {
+	type controllerCondition struct {
+		available configv1.ClusterStatusConditionType
+		degraded  configv1.ClusterStatusConditionType
+	}
+
+	// Define the controller conditions
+	controllerConditions := []controllerCondition{
+		{CapiInstallerControllerAvailableCondition, CapiInstallerControllerDegradedCondition},
+		{CoreClusterControllerAvailableCondition, CoreClusterControllerDegradedCondition},
+		{InfraClusterControllerAvailableCondition, InfraClusterControllerDegradedCondition},
+		{KubeconfigControllerAvailableCondition, KubeconfigControllerDegradedCondition},
+		{SecretSyncControllerAvailableCondition, SecretSyncControllerDegradedCondition},
+	}
+
+	if currentFeatureGates != nil && currentFeatureGates.Enabled(features.FeatureGateMachineAPIMigration) {
+		controllerConditions = append(controllerConditions, controllerCondition{MachineSyncControllerAvailableCondition, MachineSyncControllerDegradedCondition})
+		controllerConditions = append(controllerConditions, controllerCondition{MachineSetSyncControllerAvailableCondition, MachineSetSyncControllerDegradedCondition})
+	}
+
+	// Variables to store the conditions evaluation.
+	availableMsg, degradedMsg := []string{}, []string{}
+	anyAvailableMissing, anyDegradedMissing := false, false
+
+	// Evaluate each controller's conditions.
+	for _, conditions := range controllerConditions {
+		evaluateCondition(co, conditions.available, &anyAvailableMissing, &availableMsg, false)
+		evaluateCondition(co, conditions.degraded, &anyDegradedMissing, &degradedMsg, true)
+	}
+
+	availableCondition := newAvailableCondition(anyAvailableMissing, availableMsg, releaseVersion)
+	degradedCondition := newDegradedCondition(anyDegradedMissing, degradedMsg)
+	progressingCondition := newCondition(configv1.OperatorProgressing, configv1.ConditionFalse, "", "")
```

Comment:
At some point the CO should be progressing, during an upgrade of operands, I wonder why this isn't represented here?